File: packages/authentication/ios/Plugin/FirebaseAuthenticationHelper.swift
import Foundation
import Capacitor
import FirebaseCore
import FirebaseAuth

public struct ProviderId {
    static let apple = "apple.com"
    static let gameCenter = "gc.apple.com"
    static let facebook = "facebook.com"
    static let gitHub = "github.com"
    static let google = "google.com"
    static let microsoft = "microsoft.com"
    static let playGames = "playgames.google.com"
    static let twitter = "twitter.com"
    static let yahoo = "yahoo.com"
    static let password = "password"
    static let phone = "phone"
}

public class FirebaseAuthenticationHelper {
    public static func createActionCodeSettings(_ settings: JSObject?) -> ActionCodeSettings? {
        guard let settings = settings else {
            return nil
        }
        let actionCodeSettings = ActionCodeSettings()
        if let url = settings["url"] as? String {
            actionCodeSettings.url = URL(string: url)
        }
        if let handleCodeInApp = settings["handleCodeInApp"] as? Bool {
            actionCodeSettings.handleCodeInApp = handleCodeInApp
        }
        if let iOS = settings["iOS"] as? JSObject {
            if let bundleId = iOS["bundleId"] as? String {
                actionCodeSettings.setIOSBundleID(bundleId)
            }
        }
        if let android = settings["android"] as? JSObject {
            if let packageName = android["packageName"] as? String {
                actionCodeSettings.setAndroidPackageName(
                    packageName,
                    installIfNotAvailable: android["installApp"] as? Bool ?? false,
                    minimumVersion: android["minimumVersion"] as? String
                )
            }
        }
        if let linkDomain = settings["linkDomain"] as? String {
            actionCodeSettings.linkDomain = linkDomain
        }
        return actionCodeSettings
    }

    public static func createErrorCode(error: Error?) -> String? {
        if let error = error as NSError? {
            if let errorCode = convertErrorCodeToString(errorCode: error.code) {
                let prefixedErrorCode = "auth/" + errorCode
                return prefixedErrorCode
            } else {
                return nil
            }
        }
        return nil
    }

    public static func createSignInResult(credential: AuthCredential?, user: User?, idToken: String?, nonce: String?, accessToken: String?, additionalUserInfo: AdditionalUserInfo?) -> JSObject {
        return createSignInResult(credential: credential, user: user, idToken: idToken, nonce: nonce, accessToken: accessToken, additionalUserInfo: additionalUserInfo, displayName: nil)
    }

    public static func createSignInResult(credential: AuthCredential?, user: User?, idToken: String?, nonce: String?, accessToken: String?, additionalUserInfo: AdditionalUserInfo?, displayName: String?) -> JSObject {
        return createSignInResult(credential: credential, user: user, idToken: idToken, nonce: nonce, accessToken: accessToken, additionalUserInfo: additionalUserInfo, displayName: nil,
                                  authorizationCode: nil)
    }

    public static func createSignInResult(credential: AuthCredential?, user: User?, idToken: String?, nonce: String?, accessToken: String?, additionalUserInfo: AdditionalUserInfo?, displayName: String?, authorizationCode: String?) -> JSObject {
        return createSignInResult(credential: credential, user: user, idToken: idToken, nonce: nonce, accessToken: accessToken,
                                  serverAuthCode: nil, additionalUserInfo: additionalUserInfo, displayName: nil, authorizationCode: nil)
    }

    // swiftlint:disable function_parameter_count
    public static func createSignInResult(credential: AuthCredential?, user: User?, idToken: String?, nonce: String?, accessToken: String?, serverAuthCode: String?, additionalUserInfo: AdditionalUserInfo?, displayName: String?, authorizationCode: String?) -> JSObject {
        let userResult = self.createUserResult(user, displayName: displayName)
        let credentialResult = self.createCredentialResult(credential, idToken: idToken, nonce: nonce, accessToken: accessToken, authorizationCode: authorizationCode, serverAuthCode: serverAuthCode)
        let additionalUserInfoResult = self.createAdditionalUserInfoResult(additionalUserInfo)
        var result = JSObject()
        result["user"] = userResult ?? NSNull()
        result["credential"] = credentialResult ?? NSNull()
        result["additionalUserInfo"] = additionalUserInfoResult ?? NSNull()
        return result
    }

    public static func createUserResult(_ user: User?) -> JSObject? {
        return createUserResult(user, displayName: nil)
    }

    public static func createUserResult(_ user: User?, displayName: String?) -> JSObject? {
        guard let user = user else {
            if let displayName = displayName {
                var result = JSObject()
                result["displayName"] = displayName
                return result
            } else {
                return nil
            }
        }
        var result = JSObject()
        result["displayName"] = displayName ?? user.displayName ?? NSNull()
        result["email"] = user.email ?? NSNull()
        result["emailVerified"] = user.isEmailVerified
        result["isAnonymous"] = user.isAnonymous
        result["metadata"] = self.createUserMetadataResult(user.metadata)
        result["phoneNumber"] = user.phoneNumber ?? NSNull()
        result["photoUrl"] = user.photoURL?.absoluteString ?? NSNull()
        result["providerData"] = self.createUserProviderDataResult(user.providerData)
        result["providerId"] = user.providerID
        result["tenantId"] = user.tenantID ?? NSNull()
        result["uid"] = user.uid
        return result
    }

    public static func createCredentialResult(_ credential: AuthCredential?, idToken: String?, nonce: String?, accessToken: String?, authorizationCode: String?, serverAuthCode: String?) -> JSObject? {
        if credential == nil && idToken == nil && nonce == nil && accessToken == nil && authorizationCode == nil {
            return nil
        }
        var result = JSObject()
        if let credential = credential {
            result["providerId"] = credential.provider
            if let oAuthCredential = credential as? OAuthCredential {
                let oAuthAccessToken = oAuthCredential.accessToken
                if oAuthAccessToken != nil {
                    result["accessToken"] = oAuthAccessToken
                }
                let oAuthIdToken = oAuthCredential.idToken
                if oAuthIdToken != nil {
                    result["idToken"] = oAuthIdToken
                }
                let oAuthSecret = oAuthCredential.secret
                if oAuthSecret != nil {
                    result["secret"] = oAuthSecret
                }
            }
        }
        if let idToken = idToken {
            result["idToken"] = idToken
        }
        if let nonce = nonce {
            result["nonce"] = nonce
        }
        if let accessToken = accessToken {
            result["accessToken"] = accessToken
        }
        if let authorizationCode = authorizationCode {
            result["authorizationCode"] = authorizationCode
        }
        if let serverAuthCode = serverAuthCode {
            result["serverAuthCode"] = serverAuthCode
        }
        return result
    }

    public static func createAdditionalUserInfoResult(_ additionalUserInfo: AdditionalUserInfo?) -> JSObject? {
        guard let additionalUserInfo = additionalUserInfo else {
            return nil
        }
        var result = JSObject()
        result["isNewUser"] = additionalUserInfo.isNewUser
        if let profile = additionalUserInfo.profile {
            result["profile"] = JSTypes.coerceDictionaryToJSObject(profile) ?? [:]
        }
        result["providerId"] = additionalUserInfo.providerID
        if let username = additionalUserInfo.username {
            result["username"] = username
        }
        return result
    }

    private static func createUserMetadataResult(_ metadata: UserMetadata) -> JSObject {
        var result = JSObject()
        if let creationDate = metadata.creationDate?.timeIntervalSince1970 {
            result["creationTime"] = creationDate * 1000
        }
        if let lastSignInDate = metadata.lastSignInDate?.timeIntervalSince1970 {
            result["lastSignInTime"] = lastSignInDate * 1000
        }
        return result
    }

    private static func createUserProviderDataResult(_ providerData: [UserInfo]) -> JSArray {
        var result = JSArray()
        for userInfo in providerData {
            var userInfoResult = JSObject()
            userInfoResult["displayName"] = userInfo.displayName ?? NSNull()
            userInfoResult["email"] = userInfo.email ?? NSNull()
            userInfoResult["phoneNumber"] = userInfo.phoneNumber ?? NSNull()
            userInfoResult["photoUrl"] = userInfo.photoURL?.absoluteString ?? NSNull()
            userInfoResult["providerId"] = userInfo.providerID
            userInfoResult["uid"] = userInfo.uid
            result.append(userInfoResult)
        }
        return result
    }

    private static func convertErrorCodeToString(errorCode: Int) -> String? {
        let errorCodes: [Int: String] = [
            AuthErrorCode.invalidCustomToken.rawValue: "invalid-custom-token",
            AuthErrorCode.customTokenMismatch.rawValue: "custom-token-mismatch",
            AuthErrorCode.invalidCredential.rawValue: "invalid-credential",
            AuthErrorCode.userDisabled.rawValue: "user-disabled",
            AuthErrorCode.operationNotAllowed.rawValue: "operation-not-allowed",
            AuthErrorCode.emailAlreadyInUse.rawValue: "email-already-in-use",
            AuthErrorCode.invalidEmail.rawValue: "invalid-email",
            AuthErrorCode.wrongPassword.rawValue: "wrong-password",
            AuthErrorCode.tooManyRequests.rawValue: "too-many-requests",
            AuthErrorCode.userNotFound.rawValue: "user-not-found",
            AuthErrorCode.accountExistsWithDifferentCredential.rawValue: "account-exists-with-different-credential",
            AuthErrorCode.requiresRecentLogin.rawValue: "requires-recent-login",
            AuthErrorCode.providerAlreadyLinked.rawValue: "provider-already-linked",
            AuthErrorCode.noSuchProvider.rawValue: "no-such-provider",
            AuthErrorCode.invalidUserToken.rawValue: "invalid-user-token",
            AuthErrorCode.networkError.rawValue: "network-request-failed",
            AuthErrorCode.userTokenExpired.rawValue: "user-token-expired",
            AuthErrorCode.invalidAPIKey.rawValue: "invalid-api-key",
            AuthErrorCode.userMismatch.rawValue: "user-mismatch",
            AuthErrorCode.credentialAlreadyInUse.rawValue: "credential-already-in-use",
            AuthErrorCode.weakPassword.rawValue: "weak-password",
            AuthErrorCode.appNotAuthorized.rawValue: "app-not-authorized",
            AuthErrorCode.expiredActionCode.rawValue: "expired-action-code",
            AuthErrorCode.invalidActionCode.rawValue: "invalid-action-code",
            AuthErrorCode.invalidMessagePayload.rawValue: "invalid-message-payload",
            AuthErrorCode.invalidSender.rawValue: "invalid-sender",
            AuthErrorCode.invalidRecipientEmail.rawValue: "invalid-recipient-email",
            AuthErrorCode.missingEmail.rawValue: "invalid-email",
            AuthErrorCode.missingIosBundleID.rawValue: "missing-ios-bundle-id",
            AuthErrorCode.missingAndroidPackageName.rawValue: "missing-android-pkg-name",
            AuthErrorCode.unauthorizedDomain.rawValue: "unauthorized-domain",
            AuthErrorCode.invalidContinueURI.rawValue: "invalid-continue-uri",
            AuthErrorCode.missingContinueURI.rawValue: "missing-continue-uri",
            AuthErrorCode.missingPhoneNumber.rawValue: "missing-phone-number",
            AuthErrorCode.invalidPhoneNumber.rawValue: "invalid-phone-number",
            AuthErrorCode.missingVerificationCode.rawValue: "missing-verification-code",
            AuthErrorCode.invalidVerificationCode.rawValue: "invalid-verification-code",
            AuthErrorCode.missingVerificationID.rawValue: "missing-verification-id",
            AuthErrorCode.invalidVerificationID.rawValue: "invalid-verification-id",
            AuthErrorCode.missingAppCredential.rawValue: "missing-app-credential",
            AuthErrorCode.invalidAppCredential.rawValue: "invalid-app-credential",
            AuthErrorCode.sessionExpired.rawValue: "code-expired",
            AuthErrorCode.quotaExceeded.rawValue: "quota-exceeded",
            AuthErrorCode.missingAppToken.rawValue: "missing-apns-token",
            AuthErrorCode.notificationNotForwarded.rawValue: "notification-not-forwarded",
            AuthErrorCode.appNotVerified.rawValue: "app-not-verified",
            AuthErrorCode.captchaCheckFailed.rawValue: "captcha-check-failed",
            AuthErrorCode.webContextAlreadyPresented.rawValue: "cancelled-popup-request",
            AuthErrorCode.webContextCancelled.rawValue: "popup-closed-by-user",
            AuthErrorCode.appVerificationUserInteractionFailure.rawValue: "app-verification-user-interaction-failure",
            AuthErrorCode.invalidClientID.rawValue: "invalid-oauth-client-id",
            AuthErrorCode.webNetworkRequestFailed.rawValue: "network-request-failed",
            AuthErrorCode.webInternalError.rawValue: "internal-error",
            AuthErrorCode.nullUser.rawValue: "null-user",
            AuthErrorCode.keychainError.rawValue: "keychain-error",
            AuthErrorCode.internalError.rawValue: "internal-error",
            AuthErrorCode.malformedJWT.rawValue: "malformed-jwt"
        ]

        return errorCodes[errorCode]
    }
}

File: packages/authentication/android/src/main/java/app/capgo/capacitor/firebase/authentication/FirebaseAuthenticationHelper.java
package app.capgo.capacitor.firebase.authentication;

import static app.capgo.capacitor.firebase.authentication.FirebaseAuthenticationPlugin.ERROR_CODE_PREFIX;

import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.firebase.auth.ActionCodeSettings;
import com.google.firebase.auth.AdditionalUserInfo;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseAuthException;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.FirebaseUserMetadata;
import com.google.firebase.auth.OAuthCredential;
import com.google.firebase.auth.UserInfo;
import java.util.List;
import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;

public class FirebaseAuthenticationHelper {

    public static class ProviderId {

        public static final String APPLE = "apple.com";
        public static final String FACEBOOK = "facebook.com";
        public static final String GITHUB = "github.com";
        public static final String GOOGLE = "google.com";
        public static final String MICROSOFT = "microsoft.com";
        public static final String PLAY_GAMES = "playgames.google.com";
        public static final String TWITTER = "twitter.com";
        public static final String YAHOO = "yahoo.com";
        public static final String PASSWORD = "password";
        public static final String PHONE = "phone";
    }

    @Nullable
    public static ActionCodeSettings createActionCodeSettings(@Nullable JSObject settings) throws JSONException {
        if (settings == null) {
            return null;
        }
        ActionCodeSettings.Builder builder = ActionCodeSettings.newBuilder().setUrl(settings.getString("url"));
        Boolean handleCodeInApp = settings.getBoolean("handleCodeInApp", null);
        if (handleCodeInApp != null) {
            builder.setHandleCodeInApp(handleCodeInApp);
        }
        JSObject iOS = settings.getJSObject("iOS");
        if (iOS != null) {
            builder.setIOSBundleId(iOS.getString("bundleId"));
        }
        JSObject android = settings.getJSObject("android");
        if (android != null) {
            builder.setAndroidPackageName(
                android.getString("packageName"),
                android.getBoolean("installApp"),
                android.getString("minimumVersion")
            );
        }
        String dynamicLinkDomain = settings.getString("dynamicLinkDomain");
        if (dynamicLinkDomain != null) {
            builder.setDynamicLinkDomain(dynamicLinkDomain);
        }
        return builder.build();
    }

    @Nullable
    public static String createErrorCode(@Nullable Exception exception) {
        if (exception == null) {
            return null;
        } else if (exception instanceof FirebaseAuthException) {
            String errorCode = ((FirebaseAuthException) exception).getErrorCode();
            errorCode = errorCode.replaceFirst("ERROR_", "");
            String prefixedErrorCode = String.format("%s/%s", ERROR_CODE_PREFIX, errorCode);
            return snakeToKebabCase(prefixedErrorCode);
        }
        return null;
    }

    public static JSObject createSignInResult(
        @Nullable FirebaseUser user,
        @Nullable AuthCredential credential,
        @Nullable String idToken,
        @Nullable String nonce,
        @Nullable String accessToken,
        @Nullable AdditionalUserInfo additionalUserInfo
    ) {
        return FirebaseAuthenticationHelper.createSignInResult(user, credential, idToken, nonce, accessToken, null, additionalUserInfo);
    }

    public static JSObject createSignInResult(
        @Nullable FirebaseUser user,
        @Nullable AuthCredential credential,
        @Nullable String idToken,
        @Nullable String nonce,
        @Nullable String accessToken,
        @Nullable String serverAuthCode,
        @Nullable AdditionalUserInfo additionalUserInfo
    ) {
        JSObject userResult = FirebaseAuthenticationHelper.createUserResult(user);
        JSObject credentialResult = FirebaseAuthenticationHelper.createCredentialResult(
            credential,
            idToken,
            nonce,
            accessToken,
            serverAuthCode
        );
        JSObject additionalUserInfoResult = FirebaseAuthenticationHelper.createAdditionalUserInfoResult(additionalUserInfo);
        JSObject result = new JSObject();
        result.put("user", (userResult == null ? JSONObject.NULL : userResult));
        result.put("credential", (credentialResult == null ? JSONObject.NULL : credentialResult));
        result.put("additionalUserInfo", (additionalUserInfoResult == null ? JSONObject.NULL : additionalUserInfoResult));
        return result;
    }

    @Nullable
    public static JSObject createUserResult(@Nullable FirebaseUser user) {
        if (user == null) {
            return null;
        }
        JSObject result = new JSObject();
        result.put("displayName", (user.getDisplayName() == null ? JSONObject.NULL : user.getDisplayName()));
        result.put("email", (user.getEmail() == null ? JSONObject.NULL : user.getEmail()));
        result.put("emailVerified", user.isEmailVerified());
        result.put("isAnonymous", user.isAnonymous());
        result.put("metadata", FirebaseAuthenticationHelper.createUserMetadataResult(user.getMetadata()));
        result.put("phoneNumber", (user.getPhoneNumber() == null ? JSONObject.NULL : user.getPhoneNumber()));
        result.put("photoUrl", (user.getPhotoUrl() == null ? JSONObject.NULL : user.getPhotoUrl()));
        result.put("providerData", FirebaseAuthenticationHelper.createUserProviderDataResult(user.getProviderData()));
        result.put("providerId", user.getProviderId());
        result.put("tenantId", (user.getTenantId() == null ? JSONObject.NULL : user.getTenantId()));
        result.put("uid", user.getUid());
        return result;
    }

    @Nullable
    public static JSObject createCredentialResult(
        @Nullable AuthCredential credential,
        @Nullable String idToken,
        @Nullable String nonce,
        @Nullable String accessToken,
        @Nullable String serverAuthCode
    ) {
        if (credential == null && idToken == null && nonce == null && accessToken == null) {
            return null;
        }
        JSObject result = new JSObject();
        if (credential != null) {
            result.put("providerId", credential.getProvider());
            if (credential instanceof OAuthCredential) {
                String oAuthAccessToken = ((OAuthCredential) credential).getAccessToken();
                if (oAuthAccessToken != null) {
                    result.put("accessToken", oAuthAccessToken);
                }
                String oAuthIdToken = ((OAuthCredential) credential).getIdToken();
                if (oAuthIdToken != null) {
                    result.put("idToken", oAuthIdToken);
                }
                String oAuthSecret = ((OAuthCredential) credential).getSecret();
                if (oAuthSecret != null) {
                    result.put("secret", oAuthSecret);
                }
            }
        }
        if (idToken != null) {
            result.put("idToken", idToken);
        }
        if (nonce != null) {
            result.put("nonce", nonce);
        }
        if (accessToken != null) {
            result.put("accessToken", accessToken);
        }
        if (serverAuthCode != null) {
            result.put("serverAuthCode", serverAuthCode);
        }
        return result;
    }

    @Nullable
    public static JSObject createAdditionalUserInfoResult(@Nullable AdditionalUserInfo additionalUserInfo) {
        if (additionalUserInfo == null) {
            return null;
        }
        JSObject result = new JSObject();
        result.put("isNewUser", additionalUserInfo.isNewUser());
        if (additionalUserInfo.getProfile() != null) {
            JSObject profileResult = new JSObject();
            for (Map.Entry<String, Object> entry : additionalUserInfo.getProfile().entrySet()) {
                profileResult.put(entry.getKey(), entry.getValue());
            }
            result.put("profile", profileResult);
        }
        if (additionalUserInfo.getProviderId() != null) {
            result.put("providerId", additionalUserInfo.getProviderId());
        }
        if (additionalUserInfo.getUsername() != null) {
            result.put("username", additionalUserInfo.getUsername());
        }
        return result;
    }

    private static JSObject createUserMetadataResult(@Nullable FirebaseUserMetadata metadata) {
        JSObject result = new JSObject();
        if (metadata == null) {
            return result;
        }
        result.put("creationTime", metadata.getCreationTimestamp());
        result.put("lastSignInTime", metadata.getLastSignInTimestamp());
        return result;
    }

    private static JSArray createUserProviderDataResult(List<? extends UserInfo> providerData) {
        JSArray result = new JSArray();
        for (UserInfo userInfo : providerData) {
            JSObject userInfoResult = new JSObject();
            userInfoResult.put("displayName", (userInfo.getDisplayName() == null ? JSONObject.NULL : userInfo.getDisplayName()));
            userInfoResult.put("email", (userInfo.getEmail() == null ? JSONObject.NULL : userInfo.getEmail()));
            userInfoResult.put("phoneNumber", (userInfo.getPhoneNumber() == null ? JSONObject.NULL : userInfo.getPhoneNumber()));
            userInfoResult.put("photoUrl", (userInfo.getPhotoUrl() == null ? JSONObject.NULL : userInfo.getPhotoUrl()));
            userInfoResult.put("providerId", userInfo.getProviderId());
            userInfoResult.put("uid", userInfo.getUid());
            result.put(userInfoResult);
        }
        return result;
    }

    private static String snakeToKebabCase(String snakeCase) {
        return snakeCase.replaceAll("_+", "-").toLowerCase();
    }
}
